File: analysis/import/SCMLogParser/src/test/java/de/maibornwolff/codecharta/importer/scmlogparser/input/VersionControlledFileTest.java
package de.maibornwolff.codecharta.importer.scmlogparser.input;

import de.maibornwolff.codecharta.importer.scmlogparser.input.metrics.Metric;
import de.maibornwolff.codecharta.importer.scmlogparser.input.metrics.MetricsFactory;
import org.junit.Test;

import java.time.OffsetDateTime;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.Map;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.*;

public class VersionControlledFileTest {

    @Test
    public void versionControlledFileHoldsInitallyOnlyTheFilename() {
        // given
        MetricsFactory metricsFactory = mock(MetricsFactory.class);
        String filename = "filename";

        // when
        VersionControlledFile versionControlledFile = new VersionControlledFile(filename, metricsFactory);

        // then
        assertThat(versionControlledFile.getFilename()).isEqualTo(filename);
        assertThat(versionControlledFile.getActualFilename()).isEqualTo(filename);
        assertThat(versionControlledFile.getAuthors()).isEmpty();
        assertThat(versionControlledFile.markedDeleted()).isFalse();
    }


    @Test
    public void canRegisterMetricsByMetricsFactory() {
        // given
        String metricName = "metric";
        Metric metric = mock(Metric.class);
        when(metric.metricName()).thenReturn(metricName);
        when(metric.value()).thenReturn(Map.of(metricName, 1));
        when(metric.value(metricName)).thenReturn(1);

        MetricsFactory metricsFactory = mock(MetricsFactory.class);
        when(metricsFactory.createMetrics()).thenReturn(Arrays.asList(metric));

        VersionControlledFile versionControlledFile = new VersionControlledFile(
                "filename",
                metricsFactory
        );

        // when
        Map<String, ? extends Number> metricsMap = versionControlledFile.getMetricsMap();

        // then
        assertThat(metricsMap).hasSize(1);
        assertThat(versionControlledFile.getMetricValue(metricName))
                .isEqualTo(1);
    }

    @Test
    public void ignoresCommitsForDifferentFiles() {
        // given
        Metric modificationMetric = mock(Metric.class);

        VersionControlledFile versionControlledFile = new VersionControlledFile(
                "filename",
                Arrays.asList(modificationMetric)
        );

        // when
        Modification modification = new Modification("anotherFilename");
        Commit commit = createCommit("An Author", modification);
        versionControlledFile.registerCommit(commit);

        // then
        assertThat(versionControlledFile.getAuthors()).isEmpty();

        verify(modificationMetric, times(0)).registerModification(any());
    }

    @Test
    public void canRegisterASimpleCommit() {
        // given
        Metric modificationMetric = mock(Metric.class);

        String filename = "filename";
        String author = "An Author";
        VersionControlledFile versionControlledFile = new VersionControlledFile(
                filename,
                Arrays.asList(modificationMetric)
        );

        // when
        Modification modification = new Modification(filename);
        Commit commit = createCommit(author, modification);
        versionControlledFile.registerCommit(commit);

        // then
        assertThat(versionControlledFile.getFilename()).isEqualTo(filename);
        assertThat(versionControlledFile.getActualFilename()).isEqualTo(filename);
        assertThat(versionControlledFile.getAuthors()).containsExactly(author);
        assertThat(versionControlledFile.markedDeleted()).isFalse();

        verify(modificationMetric, times(1)).registerModification(any());
        verify(modificationMetric, times(1)).registerModification(eq(modification));
    }

    @Test
    public void getAuthorsReturnsAllAuthors() {
        // given
        String filename = "filename";
        String author1 = "An Author";
        String author2 = "2nd Author";
        VersionControlledFile versionControlledFile = new VersionControlledFile(filename);

        // when
        Modification modification1 = new Modification(filename);
        versionControlledFile.registerCommit(createCommit(author1, modification1));
        Modification modification2 = new Modification(filename);
        versionControlledFile.registerCommit(createCommit(author2, modification2));
        Modification modification3 = new Modification(filename);
        versionControlledFile.registerCommit(createCommit(author1, modification3));

        // then
        assertThat(versionControlledFile.getFilename()).isEqualTo(filename);
        assertThat(versionControlledFile.getAuthors()).containsExactlyInAnyOrder(author1, author2);
    }

    @Test
    public void deletionMarksFileAsDeleted() {
        // given
        String filename = "filename";
        VersionControlledFile versionControlledFile = new VersionControlledFile(filename);

        // when
        List<Modification> modifications = Arrays.asList(
                new Modification(filename),
                new Modification(filename, Modification.Type.DELETE),
                new Modification(filename)
        );
        modifications
                .forEach(
                        mod -> versionControlledFile.registerCommit(createCommit("An Author", mod))
                );

        // then
        assertThat(versionControlledFile.getFilename()).isEqualTo(filename);
        assertThat(versionControlledFile.getActualFilename()).isEqualTo(filename);
        assertThat(versionControlledFile.markedDeleted()).isTrue();
    }

    @Test
    public void renamingChangesActualFilename() {
        // given
        Metric modificationMetric = mock(Metric.class);

        String oldFilename = "old filename";
        String filename = "filename";
        VersionControlledFile versionControlledFile = new VersionControlledFile(filename,Arrays.asList(modificationMetric));

        // when
        // anti-chronological ordering
        List<Modification> modifications = Arrays.asList(
                new Modification(filename),
                new Modification(filename, oldFilename, Modification.Type.RENAME),
                new Modification(oldFilename),
                new Modification(filename)
        );
        modifications
                .forEach(
                        mod -> versionControlledFile.registerCommit(createCommit("An Author", mod))
                );

        // then
        assertThat(versionControlledFile.getFilename()).isEqualTo(oldFilename);
        assertThat(versionControlledFile.getActualFilename()).isEqualTo(filename);
        assertThat(versionControlledFile.markedDeleted()).isFalse();

        verify(modificationMetric, times(3)).registerModification(any());
    }

    private Commit createCommit(String author, Modification modification) {
        return new Commit(author, Arrays.asList(modification), OffsetDateTime.now());
    }

}

File: analysis/import/SCMLogParser/src/test/kotlin/de/maibornwolff/codecharta/importer/scmlogparser/input/VersionControlledFileTest.kt
package de.maibornwolff.codecharta.importer.scmlogparser.input

import de.maibornwolff.codecharta.importer.scmlogparser.input.metrics.Metric
import de.maibornwolff.codecharta.importer.scmlogparser.input.metrics.MetricsFactory
import io.mockk.every
import io.mockk.mockk
import io.mockk.verify
import org.assertj.core.api.Assertions.assertThat
import org.junit.Test
import org.junit.jupiter.api.Assertions
import java.time.OffsetDateTime
import java.util.* // ktlint-disable no-wildcard-imports

class VersionControlledFileTest {
    @Test
    fun versionControlledFileHoldsInitallyOnlyTheFilename() {
        // given
        val metricsFactory = mockk<MetricsFactory>(relaxed = true)
        val filename = "filename"
        // when
        val versionControlledFile = VersionControlledFile(filename, metricsFactory)
        // then
        assertThat(versionControlledFile.filename).isEqualTo(filename)
        assertThat(versionControlledFile.actualFilename).isEqualTo(filename)
        assertThat(versionControlledFile.authors).isEmpty()
        assertThat(versionControlledFile.markedDeleted()).isFalse()
    }

    @Test
    fun canRegisterMetricsByMetricsFactory() {
        // given
        val metricName = "metric"
        val metric = mockk<Metric>()
        every { metric.metricName() } returns metricName
        every { metric.value() } returns 1
        val metricsFactory = mockk<MetricsFactory>()
        every { metricsFactory.createMetrics() } returns Arrays.asList(metric)
        val versionControlledFile = VersionControlledFile(
            "filename",
            metricsFactory
        )
        // when
        val metricsMap = versionControlledFile.metricsMap
        // then
        assertThat(metricsMap).hasSize(1)
        assertThat(versionControlledFile.getMetricValue(metricName))
            .isEqualTo(1)
    }

    @Test
    fun throwsExceptionIfFileIsNotInCommit() {
        val versionControlledFile = VersionControlledFile(
            "filename",
            listOf()
        )
        val modification = Modification("anotherFilename")
        val commit = createCommit("An Author", modification)

        Assertions.assertThrows(IllegalStateException::class.java) {
            versionControlledFile.registerCommit(commit)
        }
    }

    @Test
    fun canRegisterASimpleCommit() {
        // given
        val modificationMetric = mockk<Metric>(relaxed = true)
        val filename = "filename"
        val author = "An Author"
        val versionControlledFile = VersionControlledFile(
            filename,
            Arrays.asList(modificationMetric)
        )
        // when
        val modification = Modification(filename)
        val commit = createCommit(author, modification)
        versionControlledFile.registerCommit(commit)
        // then
        assertThat(versionControlledFile.filename).isEqualTo(filename)
        assertThat(versionControlledFile.actualFilename).isEqualTo(filename)
        assertThat(versionControlledFile.authors).containsExactly(author)
        assertThat(versionControlledFile.markedDeleted()).isFalse()

        verify(exactly = 1) { modificationMetric.registerModification(any()) }
        verify(exactly = 1) { modificationMetric.registerModification(modification) }
    }

    @Test
    fun getAuthorsReturnsAllAuthors() {
        // given
        val filename = "filename"
        val author1 = "An Author"
        val author2 = "2nd Author"
        val versionControlledFile = VersionControlledFile(filename)
        // when
        val modification1 = Modification(filename)
        versionControlledFile.registerCommit(createCommit(author1, modification1))
        val modification2 = Modification(filename)
        versionControlledFile.registerCommit(createCommit(author2, modification2))
        val modification3 = Modification(filename)
        versionControlledFile.registerCommit(createCommit(author1, modification3))
        // then
        assertThat(versionControlledFile.filename).isEqualTo(filename)
        assertThat(versionControlledFile.authors).containsExactlyInAnyOrder(author1, author2)
    }

    @Test
    fun deletionMarksFileAsDeleted() {
        // given
        val filename = "filename"
        val versionControlledFile = VersionControlledFile(filename)
        // when
        val modifications = Arrays.asList(
            Modification(filename),
            Modification(filename, Modification.Type.DELETE),
            Modification(filename)
        )
        modifications
            .forEach { mod -> versionControlledFile.registerCommit(createCommit("An Author", mod)) }
        // then
        assertThat(versionControlledFile.filename).isEqualTo(filename)
        assertThat(versionControlledFile.actualFilename).isEqualTo(filename)
        assertThat(versionControlledFile.markedDeleted()).isTrue()
    }

    @Test
    fun renamingChangesActualFilename() {
        // given
        val modificationMetric = mockk<Metric>(relaxed = true)
        val oldFilename = "old filename"
        val filename = "filename"
        val versionControlledFile = VersionControlledFile(filename, Arrays.asList(modificationMetric))
        // when
        // anti-chronological ordering
        val modifications = Arrays.asList(
            Modification(filename),
            Modification(filename, oldFilename, Modification.Type.RENAME),
            Modification(oldFilename)
        )
        modifications
            .forEach { mod -> versionControlledFile.registerCommit(createCommit("An Author", mod)) }
        // then
        assertThat(versionControlledFile.filename).isEqualTo(oldFilename)
        assertThat(versionControlledFile.actualFilename).isEqualTo(filename)
        assertThat(versionControlledFile.markedDeleted()).isFalse()

        verify(exactly = 3) { modificationMetric.registerModification(any()) }
    }

    private fun createCommit(author: String, modification: Modification): Commit {
        return Commit(author, Arrays.asList(modification), OffsetDateTime.now())
    }
}